rasa/nlu/featurizers/featurizer.py

```diff
@@ -1,48 +1,142 @@
-        )
+        This name should be unique among all featurizers in the graph where this
+        component is used.
+        """
+        return self._identifier
 
+    # TODO: validate config for this component in the context of a recipe
 
-class SparseFeaturizer(Featurizer):
-    pass
+    @classmethod
+    def validate_config(cls, config: Dict[Text, Any]) -> None:
+        """Validates that the featurizer is configured properly."""
+        identifier = config.get(FEATURIZER_CLASS_ALIAS, [])
+        if not identifier:
+            raise ValueError(
+                f"Expected the config to map key `{FEATURIZER_CLASS_ALIAS}` "
+                f"to some name only found {config}"
+            )
+
+    def validate_compatibility_with_tokenizer(self, tokenizer: Tokenizer) -> None:
+        """Validate that the featurizer is compatible with the given tokenizer."""
+        # NOTE: this wasn't done before, there was just a comment
+        # TODO: add (something like) this to graph validation
+        pass
+
+    # @abstractmethod
+    # def train(self, training_data: TrainingData,) -> None:
+    #     """Trains the featurizer.
+    # NOTE: do we want a common train? For count vectorizer we then need some
+    # `setup(spacy_nlp)` - `train(spacy_nlp, training_data)` (and no common interface)
+
+    @abstractmethod
+    def process(self, messages: List[Message],) -> List[Message]:
+        """Featurizes all given messages in-place.
+
+        Args:
+          messages: messages to be featurized
+        Returns:
+          the same list with the same messages after featurization
+        """
+        ...
+
+    def process_training_data(self, training_data: TrainingData) -> TrainingData:
+        """Processes the training examples in the given training data in-place.
+
+        Args:
+          training_data: the training data
+        Returns:
+          same training data after processing
+        """
+        self.process(training_data.training_examples)
+        return training_data
+
+    def add_features_to_message(
+        self,
+        sequence: FeatureType,
+        sentence: Optional[FeatureType],
+        attribute: Text,
+        message: Message,
+    ) -> None:
+        """Adds sequence and sentence features for the attribute to the given message.
+
+        Args:
+          sequence: sequence feature matrix
+          sentence: sentence feature matrix
+          attribute: the attribute which both features describe
+          message: the message to which we want to add those features
+        """
+        for type, features in [
+            (FEATURE_TYPE_SEQUENCE, sequence),
+            (FEATURE_TYPE_SENTENCE, sentence),
+        ]:
+            if features is not None:
+                wrapped_feature = Features(features, type, attribute, self.identifier,)
+                message.add_features(wrapped_feature)
+
+    @staticmethod
+    def validate_featurizers_compatible(featurizers: Iterable[Featurizer2]) -> None:
```

Comment:
By what is this called?

Reply:
Was thinking the graph should do a static checks of the configs. All of these validate_* functions should be class functions (one is not, will fix that now) that can be used during those checks.

Reply:
I think if we do this then this should be done by the Recipes as they know which role which component has in the graph.  I'd lean towards doing all of this later and prioritizing having a minimal running version later. Once this all works we can work on adding more validations etc.

Reply:
Yep, but the validate_* logic should remain in the classes, no?